factcast-test/src/main/java/org/factcast/test/FactCastExtension.java

```diff
@@ -15,61 +15,88 @@
+        AfterAllCallback {
 
-  @Override
-  public void beforeEach(ExtensionContext extensionContext) throws Exception {
-    Class<?> testClass =
-        extensionContext
-            .getTestClass()
-            .orElseThrow(() -> new IllegalArgumentException("TestClass cannot be resolved"));
+  private static boolean initialized = false;
+  private static final List<FactCastIntegrationTestExtension> extensions = new LinkedList<>();
+  private static final Map<String, GenericContainer<?>> containers = new ConcurrentHashMap<>();
+  private static List<FactCastIntegrationTestExtension> reverseExtensions;
 
-    while (testClass.getEnclosingClass() != null) {
-      testClass = testClass.getEnclosingClass();
+  @Override
+  public void beforeEach(ExtensionContext context) throws Exception {
+    for (FactCastIntegrationTestExtension e : extensions) {
+      e.beforeEach(context);
     }
+  }
 
-    val pg = findPG(testClass);
+  @Override
+  public void afterAll(ExtensionContext context) throws Exception {
+    for (FactCastIntegrationTestExtension e : reverseExtensions) {
+      e.afterAll(context);
+    }
+  }
 
-    if (pg.isPresent()) {
-      log.debug("Wiping FactCast data from postgresql");
-      PostgresEraser.wipeAllFactCastDataDataFromPostgres(pg.get());
-    } else {
-      log.warn(
-          "No static field of type {} found, so wiping data from Postgres was not possible.",
-          PostgreSQLContainer.class.getName());
+  @Override
+  public void afterEach(ExtensionContext context) throws Exception {
+    for (FactCastIntegrationTestExtension e : reverseExtensions) {
+      e.afterEach(context);
     }
   }
 
-  private Optional<? extends PostgreSQLContainer<?>> findPG(Class<?> testClass) {
-    return org.junit.platform.commons.util.ReflectionUtils.findFields(
-            testClass,
-            FactCastExtension::isPostgresContainer,
-            ReflectionUtils.HierarchyTraversalMode.TOP_DOWN)
-        .stream()
-        .map(
-            f -> {
-              try {
-                f.setAccessible(true);
-                return (PostgreSQLContainer<?>) f.get(null);
-              } catch (IllegalAccessException e) {
-                throw new IllegalArgumentException(
-                    "Cannot get value from (supposedly static) field " + f);
-              }
-            })
-        .findAny();
+  @Override
+  public void beforeAll(ExtensionContext context) throws Exception {
+    synchronized (extensions) {
+      if (!initialized) {
+        initialize();
+        initialized = true;
+      }
+    }
+    for (FactCastIntegrationTestExtension e : extensions) {
+      e.beforeAll(context);
+    }
   }
 
-  private static boolean isPostgresContainer(Field f) {
-    return ModifierSupport.isStatic(f) && PostgreSQLContainer.class.isAssignableFrom(f.getType());
+  private void initialize() {
+    val discovered =
+        Lists.newArrayList(ServiceLoader.load(FactCastIntegrationTestExtension.class).iterator());
```

Comment:
interesting

Reply:
needed due to potentially missing dependency to the redis modules